infrastructure/src/main/java/org/corfudb/infrastructure/CompactorService.java

```diff
@@ -35,36 +35,62 @@
 public class CompactorService implements ManagementService {
 
     private final ServerContext serverContext;
-    private final SingletonResource<CorfuRuntime> runtimeSingletonResource;
-    private final ScheduledExecutorService orchestratorThread;
+    private final Duration triggerInterval;
     private final InvokeCheckpointing checkpointerJvmManager;
     private final CompactionTriggerPolicy compactionTriggerPolicy;
-
+    private ScheduledExecutorService orchestratorThread;
     private Optional<CompactorLeaderServices> optionalCompactorLeaderServices = Optional.empty();
     private Optional<CorfuStore> optionalCorfuStore = Optional.empty();
     private Optional<DistributedCheckpointerHelper> optionalDistributedCheckpointerHelper = Optional.empty();
+    private Optional<CorfuRuntime> corfuRuntimeOptional = Optional.empty();
     private TrimLog trimLog;
     private final Logger log;
     private static final Duration LIVENESS_TIMEOUT = Duration.ofMinutes(1);
+    private static final int SYSTEM_DOWN_HANDLER_TRIGGER_LIMIT = 60;
 
-    CompactorService(@NonNull ServerContext serverContext,
-                     @NonNull SingletonResource<CorfuRuntime> runtimeSingletonResource,
+    public CompactorService(@NonNull ServerContext serverContext,
+                     @NonNull Duration triggerInterval,
                      @NonNull InvokeCheckpointing checkpointerJvmManager,
                      @NonNull CompactionTriggerPolicy compactionTriggerPolicy) {
         this.serverContext = serverContext;
-        this.runtimeSingletonResource = runtimeSingletonResource;
-
-        this.orchestratorThread = Executors.newSingleThreadScheduledExecutor(
-                new ThreadFactoryBuilder()
-                        .setNameFormat("Cmpt-" + serverContext.getServerConfig().get("<port>") + "-chkpter")
-                        .build());
+        this.triggerInterval = triggerInterval;
         this.checkpointerJvmManager = checkpointerJvmManager;
         this.compactionTriggerPolicy = compactionTriggerPolicy;
         this.log = LoggerFactory.getLogger("compactor-leader");
     }
 
-    private CorfuRuntime getCorfuRuntime() {
-        return runtimeSingletonResource.get();
+    private Runnable getSystemDownHandlerForCompactor() {
+        return () -> {
+            log.warn("CorfuRuntime for CompactorService stalled. Invoking systemDownHandler after {} "
+                    + "unsuccessful tries.", SYSTEM_DOWN_HANDLER_TRIGGER_LIMIT);
+            shutdown();
+            start(this.triggerInterval);
+        };
+    }
```

Comment:
This looks like a memory leak + concurrency bug

What happens if this handler is invoked from different threads?

Reply:
Done